### PR TITLE
Resilient source

### PR DESCRIPTION
**Pull request by rozza**
Can now support non-existent collections / databases
And collection drops

KAFKA-57 KAFKA-59

### CODE REVIEW

**Comment by rozza** [comment]
https://evergreen.mongodb.com/version/5ddfad4d61837d3829dd8d47

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java [hunk 1]
```diff
@@ -140,8 +140,8 @@ public void start(final Map<String, String> props) {
 
                 // If the cursor is invalidated add the record and return calls
                 if (changeStreamDocument.getString("operationType", new BsonString("")).getValue().equalsIgnoreCase(INVALIDATE)) {
-                    LOGGER.info("Cursor has been invalidated, no further messages will be published");
-                    isRunning.set(false);
+                    LOGGER.info("Cursor has been invalidated.");
+                    cursor = null;
```
Comment (jyemin) [inline src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:144]: Not sure if we should assume that the cursor is exhausted.  Consider calling `close` first.

Comment (jyemin) [inline src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:144]: Should we be setting the resume token here to use when creating the new cursor, or is that already taken care of?

File: src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java [hunk 2]
```diff
@@ -134,10 +141,38 @@ void testSourceLoadsDataFromDatabase() {
...
+            sourceProperties.put(MongoSourceConfig.DATABASE_CONFIG, db.getName());
+            addSourceConnector(sourceProperties);
+
+            Thread.sleep(5000);
+            assertProduced(0, coll);
+
+            insertMany(rangeClosed(1, 100), coll);
+            assertProduced(100, coll);
+
+            coll.drop();
```
Comment (jyemin) [inline src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java]: Can a database drop be survived?  IIRC you mentioned something about that last week.

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java (hunk 1 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:144]: Added

Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:144]: Its already saved as part of the sourceRecord metadata

File: src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java (hunk 2 above)
Comment (rozza) [reply src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java]: Yes  - updating MongoDB seemed to fix it.

Comment (jyemin) [reply src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java]: Should this be a db.drop() then?

Comment (rozza) [reply src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java]: Good catch. I had to put a sleep in as the collection drop event happens first then the invalidate event some time after.

File: src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java (hunk 1 above)
Comment (rozza) [reply src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:144]: I reverted this change as Invalidate events already close the cursor.  https://docs.mongodb.com/manual/reference/change-events/#invalidate-event

Comment (jyemin) [reply src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:144]: Ack

File: src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java (hunk 2 above)
Comment (jyemin) [reply src/integrationTest/java/com/mongodb/kafka/connect/MongoSourceConnectorTest.java]: Unfortunate we need the sleep.  I suppose if it causes us problems we can come up with something clever to replace it.

**Review by jyemin — APPROVED** [review]
LGTM!

**Pull request closed without merging**